File: niko-studio/src/memory/query_cache.py
"""
Query Embedding Cache with LRU and TTL

Features:
- LRU cache with configurable max size (default: 1000)
- TTL-based expiration (default: 1 hour)
- Thread-safe async operations
- Memory-efficient storage
"""

import asyncio
import hashlib
import logging
import time
from collections import OrderedDict
from dataclasses import dataclass
from typing import List, Optional, Tuple

logger = logging.getLogger("niko-cache")


@dataclass
class CacheEntry:
    """Cache entry with embedding and timestamp."""
    embedding: List[float]
    created_at: float
    access_count: int = 0


class QueryEmbeddingCache:
    """
    LRU cache for query embeddings with TTL support.

    Usage:
        cache = QueryEmbeddingCache(max_size=1000, ttl_seconds=3600)

        # Try to get from cache
        embedding = cache.get(query)
        if embedding is None:
            embedding = embedder.embed(query)
            cache.put(query, embedding)
    """

    def __init__(
        self,
        max_size: int = 1000,
        ttl_seconds: float = 3600.0  # 1 hour default
    ):
        self.max_size = max_size
        self.ttl_seconds = ttl_seconds

        self._cache: OrderedDict[str, CacheEntry] = OrderedDict()
        self._lock = asyncio.Lock()

        # Statistics
        self._hits = 0
        self._misses = 0

    def _hash_query(self, query: str) -> str:
        """Generate a consistent hash for the query."""
        return hashlib.sha256(query.encode('utf-8')).hexdigest()[:16]

    def _is_expired(self, entry: CacheEntry) -> bool:
        """Check if a cache entry has expired."""
        return (time.time() - entry.created_at) > self.ttl_seconds

    def get(self, query: str) -> Optional[List[float]]:
        """
        Get embedding from cache (sync version).

        Args:
            query: The query string

        Returns:
            Embedding list if found and not expired, None otherwise
        """
        key = self._hash_query(query)

        if key not in self._cache:
            self._misses += 1
            return None

        entry = self._cache[key]

        # Check TTL
        if self._is_expired(entry):
            del self._cache[key]
            self._misses += 1
            return None

        # Move to end (most recently used)
        self._cache.move_to_end(key)
        entry.access_count += 1
        self._hits += 1

        return entry.embedding

    async def get_async(self, query: str) -> Optional[List[float]]:
        """
        Get embedding from cache (async version).

        Args:
            query: The query string

        Returns:
            Embedding list if found and not expired, None otherwise
        """
        async with self._lock:
            return self.get(query)

    def put(self, query: str, embedding: List[float]) -> None:
        """
        Store embedding in cache (sync version).

        Args:
            query: The query string
            embedding: The embedding vector
        """
        key = self._hash_query(query)

        # Evict oldest if at capacity
        while len(self._cache) >= self.max_size:
            self._cache.popitem(last=False)

        self._cache[key] = CacheEntry(
            embedding=embedding,
            created_at=time.time(),
            access_count=1
        )

    async def put_async(self, query: str, embedding: List[float]) -> None:
        """
        Store embedding in cache (async version).

        Args:
            query: The query string
            embedding: The embedding vector
        """
        async with self._lock:
            self.put(query, embedding)

    def get_or_compute(
        self,
        query: str,
        compute_fn
    ) -> List[float]:
        """
        Get from cache or compute and store.

        Args:
            query: The query string
            compute_fn: Function to compute embedding if not cached

        Returns:
            Embedding vector
        """
        cached = self.get(query)
        if cached is not None:
            return cached

        embedding = compute_fn(query)
        self.put(query, embedding)
        return embedding

    async def get_or_compute_async(
        self,
        query: str,
        compute_fn
    ) -> List[float]:
        """
        Get from cache or compute and store (async version).

        Args:
            query: The query string
            compute_fn: Async or sync function to compute embedding

        Returns:
            Embedding vector
        """
        async with self._lock:
            cached = self.get(query)
            if cached is not None:
                return cached

        # Compute outside lock to avoid blocking
        if asyncio.iscoroutinefunction(compute_fn):
            embedding = await compute_fn(query)
        else:
            embedding = compute_fn(query)

        async with self._lock:
            self.put(query, embedding)

        return embedding

    def invalidate(self, query: str) -> bool:
        """
        Remove a specific query from cache.

        Args:
            query: The query string

        Returns:
            True if entry was removed, False if not found
        """
        key = self._hash_query(query)
        if key in self._cache:
            del self._cache[key]
            return True
        return False

    def clear(self) -> int:
        """
        Clear all entries from cache.

        Returns:
            Number of entries cleared
        """
        count = len(self._cache)
        self._cache.clear()
        return count

    def cleanup_expired(self) -> int:
        """
        Remove all expired entries.

        Returns:
            Number of entries removed
        """
        now = time.time()
        expired_keys = [
            key for key, entry in self._cache.items()
            if (now - entry.created_at) > self.ttl_seconds
        ]

        for key in expired_keys:
            del self._cache[key]

        return len(expired_keys)

    @property
    def size(self) -> int:
        """Current number of entries in cache."""
        return len(self._cache)

    @property
    def hit_rate(self) -> float:
        """Cache hit rate (0.0 - 1.0)."""
        total = self._hits + self._misses
        if total == 0:
            return 0.0
        return self._hits / total

    @property
    def stats(self) -> dict:
        """Get cache statistics."""
        return {
            "size": self.size,
            "max_size": self.max_size,
            "hits": self._hits,
            "misses": self._misses,
            "hit_rate": round(self.hit_rate, 4),
            "ttl_seconds": self.ttl_seconds
        }


# Global cache instance
_global_cache: Optional[QueryEmbeddingCache] = None


def get_query_cache(
    max_size: int = 1000,
    ttl_seconds: float = 3600.0
) -> QueryEmbeddingCache:
    """
    Get or create the global query embedding cache.

    Args:
        max_size: Maximum cache entries
        ttl_seconds: TTL in seconds

    Returns:
        QueryEmbeddingCache instance
    """
    global _global_cache
    if _global_cache is None:
        _global_cache = QueryEmbeddingCache(
            max_size=max_size,
            ttl_seconds=ttl_seconds
        )
        logger.info(
            f"Query embedding cache initialized: "
            f"max_size={max_size}, ttl={ttl_seconds}s"
        )
    return _global_cache

File: src-ts/memory/query-cache.ts
/**
 * Query Embedding Cache with LRU, TTL, and per-key deduplication
 *
 * Features:
 * - LRU cache with configurable max size (default: 1000)
 * - TTL-based expiration (default: 1 hour)
 * - Per-key Promise deduplication (replaces global async-lock)
 * - Stale-while-revalidate for concurrent lookups
 * - Memory-efficient storage
 *
 * Migrated from global async-lock (bottleneck) to per-key Promise
 * caching pattern learned from maestro-flow's Semaphore design.
 */

import { createHash } from 'crypto';

/** Cache entry with embedding and timestamp. */
export interface CacheEntry {
  /** Embedding vector. */
  embedding: number[];
  /** Creation timestamp (epoch milliseconds). */
  created_at: number;
  /** Number of times accessed. */
  access_count: number;
}

/**
 * LRU cache for query embeddings with TTL and per-key deduplication.
 *
 * Usage:
 *     const cache = new QueryEmbeddingCache(max_size=1000, ttl_seconds=3600);
 *
 *     // Try to get from cache
 *     const embedding = cache.get(query);
 *     if (embedding is null) {
 *         embedding = embedder.embed(query);
 *         cache.put(query, embedding);
 *     }
 */
export class QueryEmbeddingCache {
  private max_size: number;
  private ttl_seconds: number;
  private _cache = new Map<string, CacheEntry>();
  private _hits = 0;
  private _misses = 0;

  /** Per-key pending Promise deduplication — replaces global lock. */
  private _pending = new Map<string, Promise<number[]>>();

  constructor(max_size: number = 1000, ttl_seconds: number = 3600.0) {
    this.max_size = max_size;
    this.ttl_seconds = ttl_seconds;
    this._cache = new Map();
    this._pending = new Map();
    this._hits = 0;
    this._misses = 0;
  }

  /** Generate a consistent hash for the query. */
  private _hashQuery(query: string): string {
    return createHash('sha256').update(query).digest('hex').substring(0, 16);
  }

  /** Check if a cache entry has expired. */
  private _isExpired(entry: CacheEntry): boolean {
    return (Date.now() - entry.created_at) > this.ttl_seconds;
  }

  /**
   * Get embedding from cache (sync version).
   *
   * @param query - The query string
   * @returns Embedding list if found and not expired, null otherwise
   */
  get(query: string): number[] | null {
    const key = this._hashQuery(query);

    if (!this._cache.has(key)) {
      this._misses++;
      return null;
    }

    const entry = this._cache.get(key)!;

    if (this._isExpired(entry)) {
      this._cache.delete(key);
      this._misses++;
      return null;
    }

    // Move to end (most recently used): delete then re-insert
    this._cache.delete(key);
    this._cache.set(key, entry);
    entry.access_count++;
    this._hits++;

    return entry.embedding;
  }

  /**
   * Get embedding from cache (async version).
   *
   * No global lock — uses per-key deduplication when combined
   * with getOrComputeAsync.
   */
  async getAsync(query: string): Promise<number[] | null> {
    return this.get(query);
  }

  /**
   * Store embedding in cache (sync version).
   *
   * @param query - The query string
   * @param embedding - The embedding vector
   */
  put(query: string, embedding: number[]): void {
    const key = this._hashQuery(query);

    while (this._cache.size >= this.max_size) {
      const firstKey = this._cache.keys().next().value;
      if (firstKey !== undefined) {
        this._cache.delete(firstKey);
      } else {
        break;
      }
    }

    this._cache.set(key, {
      embedding,
      created_at: Date.now(),
      access_count: 1,
    });
  }

  /**
   * Store embedding in cache (async version).
   */
  async putAsync(query: string, embedding: number[]): Promise<void> {
    this.put(query, embedding);
  }

  /**
   * Get from cache or compute and store (sync version).
   *
   * @param query - The query string
   * @param compute_fn - Function to compute embedding if not cached
   * @returns Embedding vector
   */
  getOrCompute(query: string, compute_fn: (query: string) => number[]): number[] {
    const cached = this.get(query);
    if (cached !== null) {
      return cached;
    }

    const embedding = compute_fn(query);
    this.put(query, embedding);
    return embedding;
  }

  /**
   * Get from cache or compute and store (async version).
   *
   * Uses per-key Promise deduplication instead of a global lock:
   * - Concurrent requests for the same key share one computation
   * - Different keys compute in parallel without blocking
   * - No global lock contention
   *
   * @param query - The query string
   * @param compute_fn - Async or sync function to compute embedding
   * @returns Embedding vector
   */
  async getOrComputeAsync(
    query: string,
    compute_fn: (query: string) => number[] | Promise<number[]>,
  ): Promise<number[]> {
    const cached = this.get(query);
    if (cached !== null) {
      return cached;
    }

    // Per-key deduplication: if this key is already being computed, await it
    const key = this._hashQuery(query);
    const pending = this._pending.get(key);
    if (pending) {
      return pending;
    }

    // Create the computation promise and register it
    const computePromise = (async () => {
      try {
        const maybePromise = compute_fn(query);
        const embedding = maybePromise instanceof Promise
          ? await maybePromise
          : maybePromise;
        this.put(query, embedding);
        return embedding;
      } finally {
        this._pending.delete(key);
      }
    })();

    this._pending.set(key, computePromise);
    return computePromise;
  }

  /**
   * Remove a specific query from cache.
   *
   * @param query - The query string
   * @returns True if entry was removed, False if not found
   */
  invalidate(query: string): boolean {
    const key = this._hashQuery(query);
    if (this._cache.has(key)) {
      this._cache.delete(key);
      return true;
    }
    return false;
  }

  /**
   * Clear all entries from cache.
   *
   * @returns Number of entries cleared
   */
  clear(): number {
    const count = this._cache.size;
    this._cache.clear();
    this._pending.clear();
    return count;
  }

  /**
   * Remove all expired entries.
   *
   * @returns Number of entries removed
   */
  cleanupExpired(): number {
    const now = Date.now();
    const expiredKeys: string[] = [];
    for (const [key, entry] of this._cache.entries()) {
      if ((now - entry.created_at) > this.ttl_seconds) {
        expiredKeys.push(key);
      }
    }

    for (const key of expiredKeys) {
      this._cache.delete(key);
    }

    return expiredKeys.length;
  }

  /** Current number of entries in cache. */
  get size(): number {
    return this._cache.size;
  }

  /** Number of pending computations. */
  get pendingCount(): number {
    return this._pending.size;
  }

  /** Cache hit rate (0.0 - 1.0). */
  get hitRate(): number {
    const total = this._hits + this._misses;
    if (total === 0) {
      return 0.0;
    }
    return this._hits / total;
  }

  /** Get cache statistics. */
  get stats(): Record<string, unknown> {
    return {
      size: this.size,
      max_size: this.max_size,
      hits: this._hits,
      misses: this._misses,
      hit_rate: Math.round(this.hitRate * 10000) / 10000,
      ttl_seconds: this.ttl_seconds,
      pending: this._pending.size,
    };
  }
}

/** Global cache instance */
let _globalCache: QueryEmbeddingCache | null = null;

/**
 * Get or create the global query embedding cache.
 *
 * @param max_size - Maximum cache entries
 * @param ttl_seconds - TTL in seconds
 * @returns QueryEmbeddingCache instance
 */
export function getQueryCache(
  max_size: number = 1000,
  ttl_seconds: number = 3600.0
): QueryEmbeddingCache {
  if (_globalCache === null) {
    _globalCache = new QueryEmbeddingCache(max_size, ttl_seconds);
  }
  return _globalCache;
}